src/pwm.rs

```diff
@@ -258,10 +370,59 @@ macro_rules! hal {
                         .set_bit()
                 );
 
-                unsafe { mem::MaybeUninit::uninit().assume_init() }
+                Pwm {
+                    clk: clk,
+                    _tim: PhantomData,
+                    _channels: PhantomData
+                }
+            }
+            
+            pwm_impl!(
+                $TIMX, $timX, (0, 1, 2, 3), (C1, C2, C3, C4);
+                $TIMX, $timX, (0, 1, 2), (C1, C2, C3);
+                $TIMX, $timX, (0, 1, 2), (C1, C2, C4);
+                $TIMX, $timX, (0, 1, 2), (C2, C3, C4);
+                $TIMX, $timX, (0, 1), (C1, C2);
+                $TIMX, $timX, (0, 1), (C1, C3);
+                $TIMX, $timX, (0, 1), (C1, C4);
+                $TIMX, $timX, (0, 1), (C2, C3);
+                $TIMX, $timX, (0, 1), (C2, C4);
+                $TIMX, $timX, (0, 1), (C3, C4);
+                $TIMX, $timX, (0), (C1);
+                $TIMX, $timX, (0), (C2);
+                $TIMX, $timX, (0), (C3);
+                $TIMX, $timX, (0), (C4);
+            );
+
+            impl Copy for PwmChannel<$TIMX, C1> {}
+            impl Clone for PwmChannel<$TIMX, C1> {
+                fn clone(&self) -> Self {
+                    unsafe { mem::MaybeUninit::uninit().assume_init() }
+                }            
+            }
+
+            impl Copy for PwmChannel<$TIMX, C2> {}
+            impl Clone for PwmChannel<$TIMX, C2> {
+                fn clone(&self) -> Self {
+                    unsafe { mem::MaybeUninit::uninit().assume_init() }
+                }            
+            }
+
+            impl Copy for PwmChannel<$TIMX, C3> {}
+            impl Clone for PwmChannel<$TIMX, C3> {
+                fn clone(&self) -> Self {
+                    unsafe { mem::MaybeUninit::uninit().assume_init() }
+                }            
+            }
+
+            impl Copy for PwmChannel<$TIMX, C4> {}
+            impl Clone for PwmChannel<$TIMX, C4> {
+                fn clone(&self) -> Self {
+                    unsafe { mem::MaybeUninit::uninit().assume_init() }
+                }            
```

Comment:
I'm pretty sure that this is unsound. By cloning a channel, you can pass the same channel to two different places at which point `set_duty` would affect both clones.

Reply:
This is true.  But, this is a hardware device and the channel maps directly to a physical device.  If an end user clones it he or she must know that if the hardware changes it will affect the software processes referring to those devices.

Ok, with that out of the way, do you know how to do this such that there is ownership?  I would be more than happy to change this.

Reply:
I think you want `Pwm::into_channels(self) -> PWMCHANNELS`. That would allow to separate the channels and then use them independently.

Reply:
>  If an end user clones it he or she must know that if the hardware changes it will affect the software processes referring to those devices.

This goes against the idea of this crate and most of the rust embedded ecosystem. The idea is to use rusts ownership model to make sure that there is never any action at a distance.

An example of situations that this prevents (pseudocode which doesn't really use this API, it's late so I can't be bothered to make i correct. It should demonstrate the issue though):

```rust
let pwm = // pwm...;
let (c1, c2) = pwm.channels();

// Some independent driver that uses c1 and relies on a specific period
run_driver(c1, ...);

// Some other driver that uses the Pwm trait to change the period, thus affecting both c1 and c2
run_driver2(c2, pwm);

// By adding the second driver, we broke the behaviour of the first which will be a pain to debug, and debugging it would require knowledge of both driver1, driver2 and the interraction between them.
```

> Ok, with that out of the way, do you know how to do this such that there is ownership? I would be more than happy to change this.

I could be wrong about this, but I presume the reason you need copy is that otherwise you can't call `set_period` on the pwm struct after distributing the channels elsewhere.

If that is the case, I can't think of a super neat way to do things off the top of my head. I still think the thing we discussed in discord where in order to set the period, you need to guarantee that the pins are not being used elsewhere (you can give mutable references) is the best option.

Unfortunately, that makes the API a lot less inconvenient, and I suspect that implementing everything in `Pwm` might be difficult.

Hopefully my sleepy ramblings provided some insight. If something is unclear, I'll hopefully be able to give a better answer tomorrow when it's less late ;)

Reply:
I will implement without the Clone/Copy and use the `into_channels(&self)` approach.  I was just trying to maintain the current API.  I also like the Enum approach as well.  I will look to try and implement that as well.

Reply:
That's really `into_channels(self)` without `&`

Reply:
@TeXitoi I am going to close this "sub-conversation" and ask for a re-review because of the large changes that were put in by @burbull